### ISSUE TITLE
Running yarn add in a workspace break the lockfile

### ISSUE THREAD

**Issue by arcanis** [yarnpkg/yarn#4028]
```bash
rm -rf foo

(
    mkdir -p foo/a && cd foo/a
    echo '{"name":"a", "version":"1.0.0"}' > package.json
)

(
    mkdir -p foo/b && cd foo/b
    echo '{"name":"b", "version":"1.0.0"}' > package.json
)

(
    mkdir -p foo/root && cd foo/root
    echo '{"private": true, "workspaces": ["workspace"], "dependencies": {"a":"file:../a"}}' > package.json
    echo 'workspaces-experimental true' > .yarnrc
)

(
    mkdir -p foo/root/workspace && cd foo/root/workspace
    echo '{"name":"workspace", "version":"1.0.0"}' > package.json
)

(
    cd foo/root
    yarn
    cp yarn.lock good-yarn.lock
)

(
    cd foo/root/workspace
    yarn add b@file:../../b
)

(
    cd foo/root
    diff good-yarn.lock yarn.lock
)
```

**Comment by arcanis** [comment]
Note that this is not strictly related to `yarn add` - you have the same behavior if you manually add the package to the `package.json` then run `yarn install`.

**Comment by edeustace** [comment]
I have a question on the expected output of the script above now that the fix has been applied. 

I'm running into issues with relative path dependencies w/ yarn and I think the pr for this issue: #4030 is related. Before trying to describe the issue and look for a fix, I want to understand what the correct behaviour is for the script above.

Before #4030 was applied, when you ran the script above: 
* `b` was added to `root/workspace/package.json` as a dependency as `file:../../b`.
* `b` was installed to `root/node_modules`
* `b` was added as a dependency in `root/yarn.lock` as `file:../../b` (which is the wrong relative path) and `a` was removed (which it shouldn't have been).

Once the fix was applied (I'm running 7fcf4dd4d95cdfa8856cea8a5daa29d48eb2d70b): 
* `yarn add b@file:../../b` fails with an error: `error Package "b" refers to a non-existing file '"/path/to/dir/post-fix/b"'.`
* `root/workspace/package.json` is untouched
* `root/yarn.lock` is untouched

I would have thought that the correct behaviour would have been: 
* `b` was added to `root/workspace/package.json` as a dependency as `file:../../b`.
* `b` was installed to `root/node_modules`
* `b` was added as a dependency in `root/yarn.lock` as `file:../b` (note the relative path is correct now) and `a` is left intact.

I see in the [docs](https://yarnpkg.com/lang/en/docs/workspaces/#toc-limitations-caveats) that workspaces must be children of the workspace root, but there is no mention of defining a relative path dependency from a workspace package out of the workspace dir. @arcanis are you the correct person to ask about this? Thanks.
